src/analysis/bits-bounds-transfer-function.h

```diff
@@ -0,0 +1,367 @@
+        if (right.constVal.has_value()) {
+          currElement.setUpperBound(
+            std::min(Index(64),
+                     Bits::getEffectiveShifts(right.constVal.value().geti64(),
+                                              Type::i64) +
+                       left.upperBound));
+        } else {
+          currElement.setUpperBound(64);
+        }
+        break;
+      }
+      case ShrUInt64: {
+        if (right.constVal.has_value()) {
+          auto shifts = std::min(Index(Bits::getEffectiveShifts(
+                                   right.constVal.value().geti64(), Type::i64)),
+                                 left.upperBound);
+          currElement.setUpperBound(
+            std::max(Index(0), left.upperBound - shifts));
+        } else {
+          currElement.setUpperBound(64);
+        }
+        break;
+      }
+      case ShrSInt64: {
+        if (right.constVal.has_value()) {
+          if (left.upperBound == 64) {
+            currElement.setUpperBound(64);
+          } else {
+            auto shifts =
+              std::min(Index(Bits::getEffectiveShifts(
+                         right.constVal.value().geti64(), Type::i64)),
+                       left.upperBound);
+            currElement.setUpperBound(
+              std::max(Index(0), left.upperBound - shifts));
+          }
+        } else {
+          currElement.setUpperBound(64);
+        }
+        break;
+      }
+      default: {
+        addInformation = false;
+      }
+    }
+
+    if (collectingResults && addInformation) {
+      exprMaxBounds[curr] = currElement.upperBound;
+    }
+
+    currState->push(std::move(currElement));
+  }
+
+  void visitUnary(Unary* curr) {
+    MaxBitsLattice::Element unaryVal = currState->pop();
+    MaxBitsLattice::Element currElement = bitsLattice.getBottom();
+
+    bool addInformation = true;
+    switch (curr->op) {
+      case ClzInt32:
+      case CtzInt32:
+      case PopcntInt32: {
+        currElement.setUpperBound(6);
+        break;
+      }
+      case ClzInt64:
+      case CtzInt64:
+      case PopcntInt64: {
+        currElement.setUpperBound(7);
+        break;
+      }
+      case WrapInt64:
+      case ExtendUInt32: {
+        currElement.setUpperBound(unaryVal.upperBound);
+        break;
+      }
+      case ExtendS8Int32: {
+        currElement.setUpperBound(
+          unaryVal.upperBound >= 8 ? Index(32) : unaryVal.upperBound);
+        break;
+      }
+      case ExtendS16Int32: {
+        currElement.setUpperBound(
+          unaryVal.upperBound >= 16 ? Index(32) : unaryVal.upperBound);
+        break;
+      }
+      case ExtendS8Int64: {
+        currElement.setUpperBound(
+          unaryVal.upperBound >= 8 ? Index(64) : unaryVal.upperBound);
+        break;
+      }
+      case ExtendS16Int64: {
+        currElement.setUpperBound(
+          unaryVal.upperBound >= 16 ? Index(64) : unaryVal.upperBound);
+        break;
+      }
+      case ExtendS32Int64:
+      case ExtendSInt32: {
```

Comment:
What's the difference between these two instructions? I'm surprised they get the same logic.

Reply:
They get the same logic in `bits.h`.